File: cs/clean-code/Samples/PathFinder_Refactored.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;

namespace Samples
{
    public class PathFinder_Refactored
    {
        public static Point GetNextStepToTarget(Point source, Point target, IMaze maze)
        {
            var queue = new Queue<Point>();
            var used = new HashSet<Point>();
            queue.Enqueue(target);
            used.Add(target);
            while (queue.Any())
            {
                var p = queue.Dequeue();
                foreach (var neighbour in GetNeighbours(p, maze))
                {
                    if (used.Contains(neighbour)) continue;
                    if (neighbour == source)
                        return p;
                    queue.Enqueue(neighbour);
                    used.Add(neighbour);
                }
            }
            return source;
        }

        public static IMaze GenerateRandomWorld()
        {
            throw new NotImplementedException("TODO");
        }

        private static IEnumerable<Point> GetNeighbours(Point from, IMaze maze)
        {
            return new[] {new Size(1, 0), new Size(-1, 0), new Size(0, 1), new Size(0, -1)}
                .Select(shift => from + shift)
                .Where(maze.InsideMaze)
                .Where(maze.IsFree);
        }
    }
}

File: java/src/main/java/ru/kontur/courses/samples/pathfinder/PathFinderRefactored.java
package ru.kontur.courses.samples.pathfinder;

import java.awt.*;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;

public class PathFinderRefactored {
    public static Point getNextStepToTarget(Point source, Point target, Maze maze) {
        var queue = new LinkedList<Point>();
        var used = new HashSet<Point>();
        queue.add(target);
        used.add(target);
        while (!queue.isEmpty()) {
            var p = queue.poll();

            for (var neighbour : getNeighbours(p, maze)) {
                if (used.contains(neighbour)) continue;
                if (neighbour == source)
                    return p;
                queue.add(neighbour);
                used.add(neighbour);
            }
        }
        return source;
    }

    public static Maze generateRandomWorld() {
        throw new UnsupportedOperationException("TODO");
    }

    private static Iterable<Point> getNeighbours(Point from, Maze maze) {
        return Arrays.stream(
                new Point[]{
                        new Point(1, 0),
                        new Point(-1, 0),
                        new Point(0, 1),
                        new Point(0, -1)
                })
                .map(it -> new Point(from.x + it.x, from.y + it.y))
                .filter(maze::insideMaze)
                .filter(maze::isFree)
                .toList();
    }
}
